src/planner/planners/MatchSolver.cpp

```diff
@@ -45,10 +47,68 @@ Status MatchSolver::buildReturn(MatchAstContext* matchCtx, SubPlan& subPlan) {
         }
     }
 
-    auto *project = Project::make(matchCtx->qctx, subPlan.root, yields);
-    project->setInputVar(subPlan.root->outputVar());
+    auto *project = Project::make(matchCtx->qctx, current, yields);
+    project->setInputVar(current->outputVar());
     project->setColNames(std::move(colNames));
-    subPlan.root = project;
+
+    current = project;
+    if (sentence->ret()->isDistinct()) {
+        auto *dedup = Dedup::make(matchCtx->qctx, current);
+        dedup->setInputVar(current->outputVar());
+        dedup->setColNames(current->colNames());
+        current = dedup;
+    }
+
+    if (sentence->ret()->orderFactors() != nullptr) {
+        auto inputColList = current->colNames();
+        std::unordered_map<std::string, size_t> inputColIndices;
+        for (auto i = 0u; i < inputColList.size(); i++) {
+            if (!inputColIndices.emplace(inputColList[i], i).second) {
+                return Status::SemanticError("Duplicated columns not allowed: %s",
+                        inputColList[i].c_str());
+            }
+        }
+
+        std::vector<std::pair<size_t, OrderFactor::OrderType>> indexedFactors;
+        auto *factors = sentence->ret()->orderFactors();
+        for (auto &factor : factors->factors()) {
+            if (factor->expr()->kind() != Expression::Kind::kLabel) {
+                return Status::SemanticError("Only column name can be used as sort item");
+            }
+            auto *name = static_cast<const LabelExpression*>(factor->expr())->name();
+            auto iter = inputColIndices.find(*name);
+            if (iter == inputColIndices.end()) {
+                return Status::SemanticError("Column `%s' not found", name->c_str());
+            }
+            indexedFactors.emplace_back(iter->second, factor->orderType());
+        }
```

Comment:
These could be part of semantic analysis.

Reply:
Yes. But then I have to store some states somewhere so that I can access them later. So far the only place is the MatchValidator instance. I think that will make the later refactoring more tedious. We can extract them out when we add a separate context for each clause.

Reply:
We could use `MatchAstContext` for now.